src/java/org/apache/cassandra/db/monitoring/SaiSlowLog.java

```diff
@@ -0,0 +1,193 @@
+
+    private final AtomicInteger numSlowestQueries = new AtomicInteger();
+    // a structure that keeps the N slowest queries - don't use getter/setter so that it's not exposed over JMX
+    private final SaiSlowestQueriesQueue slowestQueries;
+    // KATE: make this false by default and decide on yaml property, -D flag, or JMX to enable; hook setEnabled() to change
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
+
+    static final Comparator<SlowSaiQuery> SLOW_SAI_QUERY_COMPARATOR = (SlowSaiQuery o1, SlowSaiQuery o2) -> Long.compare(o2.getDuration(), o1.getDuration());
+
+    @VisibleForTesting
+    static SaiSlowLog instance = new SaiSlowLog();
+
+    public SaiSlowLog()
+    {
+        numSlowestQueries.set(DatabaseDescriptor.getSaiSlowLogNumSlowestQueries());
+        slowestQueries = new SaiSlowestQueriesQueue(numSlowestQueries.get());
+    }
+
+    public boolean isEnabled()
+    {
+        return enabled.get();
+    }
+
+    public synchronized void setEnabled(boolean newValue)
+    {
+        boolean oldValue = this.enabled.get();
+
+        if (oldValue != newValue)
+        {
+            this.enabled.set(newValue);
+            logger.info("SAI Slow Log is now {}", newValue ? "enabled" : "disabled");
+        }
+    }
+
+    // KATE below two will be used to set and get the number of slowest queries; make it configurable through property/nodetool?
+    public int getNumSlowestQueries()
+    {
+        return numSlowestQueries.get();
+    }
+
+    public void setNumSlowestQueries(int numSlowestQueries)
+    {
+        int oldValue = this.numSlowestQueries.get();
+        if (oldValue != numSlowestQueries)
+        {
+            if (numSlowestQueries < 1)
+            {
+                throw new IllegalArgumentException("sai_slow_log_num_slowest_queries must be > 0");
+            }
+            this.numSlowestQueries.set(numSlowestQueries);
+            resizeMinMaxHeap();
+            logger.info("sai_slow_log_num_slowest_queries set to {} (was {}).", numSlowestQueries, oldValue);
+        }
+    }
+
+    private void resizeMinMaxHeap()
+    {
+        slowestQueries.resize(numSlowestQueries.get());
+    }
+
+    public List<SlowSaiQuery> retrieveRecentSlowestCqlQueries()
+    {
+        List<SlowSaiQuery> queries = slowestQueries.getAndReset();
+        // there is no ordering guarantee and we want the slowest queries first
+        queries.sort(SLOW_SAI_QUERY_COMPARATOR);
+        return queries;
+    }
+
+    /**
+     * Check if we should record a given operation in the SAI slow log. If so, extract
+     * the necessary detail from the query context and log it.
+     * @param queryContext operation to log
+     */
+    public static void maybeRecord(QueryContext queryContext)
+    {
+        if (!instance.isEnabled())
+            return;
+
+        // KATE: Remove below line to avoid logging every query; it is used just during development/testing
+        logger.debug("Recording slow SAI query");
```

Comment:
also in CNDB the logger is by default configured to run at DEBUG level, if you want to not print something by default you have to use TRACE

Reply:
Yeah, it is an option but in this case I don't think I have to keep this one. It was really just to see what I am doing in the logs as I had some bug